app/lib/report_helper.rb

```diff
@@ -37,150 +37,147 @@ def view_level_name2int(view_level)
-    total_reads = report.pipeline_output.total_reads
+  def select_taxons(report)
     pipeline_output_id = report.pipeline_output.id
-    data = TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, count_type from taxon_counts where pipeline_output_id = #{pipeline_output_id}").to_hash
+    total_reads = report.pipeline_output.total_reads
+    background_id = report.background.id
+    # Note: stdev is never 0
+    TaxonCount.select("
+      taxon_counts.tax_id         AS  tax_id,
+      taxon_counts.count_type     AS  count_type,
+      taxon_counts.tax_level      AS  tax_level,
+      taxon_lineages.genus_taxid  AS  genus_taxid,
+      taxon_names.name            AS  name,
+      tgn.name                    AS  genus_name,
+      tcn.name                    AS  category_name,
+      taxon_counts.count          AS  count,
+      (count / #{total_reads}.0
+        * 1000000.0)              AS  rpm,
+      IF(
+        stdev IS NOT NULL,
+        GREATEST(-99, LEAST(99, (((count / #{total_reads}.0 * 1000000.0) - mean) / stdev))),
+        100
+      )                           AS  zscore
+    ").joins("
+      LEFT OUTER JOIN taxon_summaries ON
+        taxon_counts.tax_id     = taxon_summaries.tax_id      AND
+        taxon_counts.count_type = taxon_summaries.count_type  AND
+        taxon_counts.tax_level  = taxon_summaries.tax_level
+    ").joins("
+      LEFT OUTER JOIN taxon_lineages ON
+        taxon_counts.tax_id = taxon_lineages.taxid
+    ").joins("
+      LEFT OUTER JOIN taxon_names ON
+        taxon_names.taxid = taxon_counts.tax_id
+    ").joins("
+      LEFT OUTER JOIN taxon_names AS tgn ON
+        tgn.taxid = taxon_lineages.genus_taxid
+    ").joins("
+      LEFT OUTER JOIN taxon_names AS tcn ON
+        tcn.taxid = taxon_lineages.superkingdom_taxid
+    ").where(
+      pipeline_output_id: pipeline_output_id,
+      taxon_summaries: {
+        background_id: background_id
+      }
+    ).to_a.map(&:attributes)
+  end
 
-    # pad cases where only one of NT/NR is present with zeroes
-    taxid_counttype = data.group_by { |h| [h["tax_id"], h["count_type"]] }
-    tax_id_set = Set.new(data.map { |h| h["tax_id"] })
-    tax_id_set.each do |taxid|
-      nt_ele = taxid_counttype[[taxid, 'NT']]
-      nr_ele = taxid_counttype[[taxid, 'NR']]
-      if nt_ele && !nr_ele
-        nr_ele_new = nt_ele[0].clone
-        nr_ele_new["count_type"] = 'NR'
-        nr_ele_new["count"] = 0
-        data << nr_ele_new
-      elsif !nt_ele && nr_ele
-        nt_ele_new = nr_ele[0].clone
-        nt_ele_new["count_type"] = 'NT'
-        nt_ele_new["count"] = 0
-        data << nt_ele_new
-      end
-    end
+  def flip_type(t)
+    t == 'NR' ? 'NT' : 'NR'
+  end
+
+  def zero_twin(taxon)
+    result = taxon.clone
+    result['count'] = 0
+    result['rpm'] = 0
+    result['count_type'] = flip_type(result['count_type'])
+    result['zscore'] = -100
```

Comment:
can you document the use of -100 here. even better define constant somewhere.

Reply:
good idea, will define z score special constants